frontends/smvparser.y

```diff
@@ -371,46 +575,64 @@ exponential_number: integer_val exponential_prefix "-" integer_val{  $$ =  $1 +
 | float_number exponential_prefix "-" integer_val {  $$ =  $1 + $2 + "-" + $4; };
 
 range_constant: integer_val TO integer_val{
-  throw PonoException("No range constant now");
+  throw PonoException("Range constants are not yet supported");
 };
 basic_expr: simple_expr{ $$ = $1;}
           | next_expr{ $$ = $1;}
 
 simple_expr: constant {
-            $$ = $1;
-}
+            $$ = $1; 
+          }
             | complex_identifier {
+            if(enc.module_flat){
               smt::Term tok = enc.terms_.at($1);
               if (enc.unsignedbv_.find($1) != enc.unsignedbv_.end() ) {
                 $$ = new SMVnode(tok, SMVnode::Unsigned);
               } else if(enc.signedbv_.find($1) != enc.signedbv_.end()){
                 $$ = new SMVnode(tok, SMVnode::Signed);
-              } else{
-                if(tok->get_sort()->get_sort_kind() == smt::BV) $$ = new SMVnode(tok,SMVnode::BV);
-                else if(tok -> get_sort()->get_sort_kind() == smt::BOOL) $$ = new SMVnode(tok,SMVnode::Boolean);
+              } else if(enc.arrayty_.find($1) != enc.arrayty_.end()){
+                $$ = new SMVnode(tok, SMVnode::WordArray);
+              }else{
+                smt::SortKind kind_ = tok->get_sort()->get_sort_kind();
+                if (kind_ == smt::BV || kind_ == smt::BOOL) $$ = new SMVnode(tok,SMVnode::Boolean);
+                else if (kind_ == smt::INT) $$ = new SMVnode(tok,SMVnode::Integer);
+                else if (kind_ == smt::REAL) $$ = new SMVnode(tok,SMVnode::Real);
+                else throw PonoException("The type of the identifier is wrong");
+              }
+            }else{
+              $$ = new identifier($1);
               }
             }
             | "(" basic_expr ")"{
+              if(enc.module_flat){
               $$ = $2;
+              }else{
+              $$ = new par_expr($2);
+              }
             }
             | OP_NOT basic_expr {
+            if(enc.module_flat){
               SMVnode *a = $2;
               SMVnode::Type bvs_a = a->getType();
-              if(bvs_a != SMVnode::BV && bvs_a != SMVnode::Unsigned && bvs_a != SMVnode::Signed && bvs_a != SMVnode::Boolean){
+              if(bvs_a != SMVnode::Unsigned && bvs_a != SMVnode::Signed && bvs_a != SMVnode::Boolean){
                 throw PonoException("Type system violation");
               }
               smt::Term e = enc.solver_->make_term(smt::BVNot, a->getTerm());
```

Comment:
This should be `Not`, not `BVNot`. I think this is also because it works with Boolector but not with other solvers. You can force the binary to use mathsat by running it with `-e interp`. Everything should still work in that case. I found this by running pono like this from the build directory: `./pono -e interp -v 1 ../samples/simple_counter.smv`